KAFKA-6299. Fix AdminClient error handling when metadata changes

cmccabe opened the pull request:
AdminClient should only call Metadata#requestUpdate when needed.

When AdminClient gets a NOT_CONTROLLER error, it should refresh its metadata and retry the request, rather than making the end-user deal with NotControllerException.

Move AdminClient's metadata management outside of NetworkClient and into AdminMetadataManager.  This will make it easier to do more sophisticated metadata management in the future, such as implementing a NodeProvider which fetches the leaders for topics.

Rather than manipulating newCalls directly, the AdminClient service thread now drains it directly into pendingCalls.  This minimizes the amount of locking we have to do, since pendingCalls is only accessed from the service thread.

tedyu commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -670,29 +674,63 @@ int nextTimeoutMs() {
             }
-            if (prevMetadataVersion != null) {
-                log.trace("Metadata is now ready.");
+            if (metadataAuthException != null) {
```
Should this if block be placed above the if block on line 729 ?

cmccabe replied:
Hmm-- I don't think so.  We don't want to return the exception of the stale metadata, if new metadata has been requested.

tedyu commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -953,6 +961,21 @@ private synchronized boolean threadShouldExit(long now, long curHardShutdownTime
+                }
+                newCalls.clear();
+            }
+            for (List<Call> calls : callsToSend.values()) {
```
Probably add comment on why synchronization on callsToSend is not needed.

cmccabe replied:
I will add a comment to the declaration of callsToSend, since this is used in many other places.

cmccabe commented:
Test failure was `org.apache.kafka.common.security.authenticator.ClientAuthenticationFailureTest.testAdminClientWithInvalidCredentials`, which is not related.

hachikuji commented:
@cmccabe The test failure does seem caused by this patch. I reproduced locally on this branch. Can you check again?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -662,40 +668,86 @@ int nextTimeoutMs() {
-         * @param prevMetadataVersion       The previous metadata version which wasn't usable.
-         * @return                          null if the metadata is usable; the current metadata
-         *                                  version otherwise
+         * @return                  null, if the metadata is usable.
```
This is kind of a weird contract. Maybe the name should be `checkMetadataFetchError` or something like that?

cmccabe replied:
checkMetadataError might be a better name

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -695,29 +699,63 @@ int nextTimeoutMs() {
+            if (shouldRequestMetadata.getAndSet(false) ||
+                    metadata.fetch().isBootstrapConfigured()) {
+                if (lowestValidMetadataVersion == lastSeenMetadataVersion) {
+                    lastSeenMetadataVersion = metadata.requestUpdate();
```
Is this necessary? It seems like the only way it's possible for the `if` condition to be true is if we already updated `lastSeenMetadata` below.

cmccabe replied:
The point of this if statement is that we shouldn't keep calling Metadata#requestUpdate over and over.  We should call it once, wait for an update, and then call it again after that if needed.

hachikuji replied:
I was not questioning the need of the `if`, but the reassignment to `lastSeenMetadataVersion`, which seemed redundant given how it is updated below. I may have answered my own question however. The problem is that the version in `Metadata` is updated in two additional cases: 1) on initial bootstrapping, and 2) after a periodic refresh. We handle the first case I think because of this reassignment and the fact that the two fields are initialized to 0.

I'm not sure about the second case, however. Say, for example, that `lowestValidMetadataVersion` and `lastSeenMetadataVersion` are both 5 and no update has been requested. After the metadata max age expires, we'll automatically trigger an update and bump `lastSeenMetadataVersion` to 6. Now if we request a metadata update, nothing will happen because we'll never reach equality again. We may be able to fix it by changing to an inequality, but I'm not sure. Some testing would be helpful.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -1971,6 +2017,13 @@ public CreatePartitionsResult createPartitions(Map<String, NewPartitions> newPar
+                // Check for controller change
+                for (ApiError error : response.errors().values()) {
+                    if (error.error() == Errors.NOT_CONTROLLER) {
+                        runnable.shouldRequestMetadata.set(true);
```
nit: not that big of a deal, but since we do the same thing in several places, maybe we could add a `requestMetadataUpdate` method to the runnable.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -1302,6 +1341,13 @@ public DescribeTopicsResult describeTopics(final Collection<String> topicNames,
                 MetadataResponse response = (MetadataResponse) abstractResponse;
+                // Check for controller change
+                for (Errors error : response.errors().values()) {
+                    if (error == Errors.NOT_CONTROLLER) {
```
Is the `NOT_CONTROLLER` error possible for Metadata requests? Also, are there any other errors we care about at this level (e.g. NOT_LEADER errors)?

cmccabe replied:
No, it's not.  Good catch.

"Not leader for partition" is something we care about for certain calls, but handling it will be harder.  I want to do that in a follow-on change.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -1139,6 +1164,13 @@ public CreateTopicsResult createTopics(final Collection<NewTopic> newTopics,
+                for (ApiError error : response.errors().values()) {
+                    if (error.error() == Errors.NOT_CONTROLLER) {
+                        runnable.shouldRequestMetadata.set(true);
+                        throw error.exception();
```
I guess I would have expected that these transient errors would get retried internally. Is the intent to do this separately?

cmccabe replied:
Good question.  Unfortunately, they can't be retried internally because we have no generic top level error code.  So if createTopics fails, you get back a response that has {foo : NOT_CONTROLLER, bar: NOT_CONTROLLER, baz: NOT_CONTROLLER, etc.}  You can't parse this without knowing the subtype, which means it has to be done here.

hachikuji replied:
To clarify my question, does the user request fail because of a transient NOT_CONTROLLER error or do we retry the request? I had expected we would retry, but perhaps we are leaving that for future work?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/Metadata.java:
```diff
@@ -277,6 +277,7 @@ else if (expireMs <= now) {
     public synchronized void failedUpdate(long now, AuthenticationException authenticationException) {
         this.lastRefreshMs = now;
         this.authenticationException = authenticationException;
+        this.version += 1;
```
Have you verified that this works for all uses of version at the moment? It seems like a minor change in behavior since we use `failedUpdate` in `DefaultMetadataUpdater` if there are no nodes in the response. Currently this would cause uses such as `ConsumerNetworkClient.awaitMetadataUpdate()` to retry before returning.

cmccabe replied:
I don't think retrying is what we want, though.  If you get an authentication exception from any broker, your auth is bad and you should fail.  It's not a retryable exception, conceptually or in terms of code

hachikuji replied:
The case I am referring to is in `DefaultMetadataupdater.handleCompletedMetadataResponse`, which is different from the authentication failure path. We have a check to ensure that the metadata response contains at least one broker. With this change, we will update the metadata version even in the case that it is empty, which will cause methods like `ConsumerNetworkClient.awaitMetadataUpdate()` to return earlier than they currently do.

cmccabe replied:
Hmm.  Maybe I'm misinterpreting, but are you suggesting that `ConsumerNetworkClient#awaitMetadataUpdate()` should wait forever (or until the timeout hits) when there is an authentication error fetching the metadata?  That doesn't seem right.

hachikuji replied:
I am definitely not suggesting that. I am not talking about authentication failures at all. I am referring to the call to `Metadata.failedUpdate` that is in `DefaultMetadataupdater.handleCompletedMetadataResponse`. With this change, the version will be incremented following an "empty" metadata update. This will cause us to exit the loop in `ConsumerNetworkClient#awaitMetadataUpdate()` even though we have not received a valid update. 

My point more generally is that we are changing the meaning of the version field inside `Metadata`. It seems we were intentionally using this before to indicate when we had received a valid new version of the metadata. But now we bump it even when there's a failure which means we can no longer use it to tell when we've seen a successful update. As far as I can tell, the impact may be minor, but we should consider all of the usages to be sure of it.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/errors/InvalidMetadataException.java:
```diff
@@ -19,7 +19,7 @@
  * An exception that may indicate the client's metadata is out of date
  */
-public abstract class InvalidMetadataException extends RetriableException {
+public class InvalidMetadataException extends RetriableException {
```
Seems this was not really necessary since we just needed a sentinel value? Note that `common/errors` is a public package.

cmccabe replied:
I can use NetworkException instead (which is a subclass of InvalidMetadataException)

hachikuji reviewed (commented):
Thanks for the patch. Left a few comments/questions.

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -209,7 +214,24 @@ public void testTimeoutWithoutMetadata() throws Exception {
+            env.kafkaClient().setNode(env.cluster().nodeById(0));
+            env.kafkaClient().prepareResponse(new CreateTopicsResponse(Collections.singletonMap("myTopic", new ApiError(Errors.NONE, ""))));
+            KafkaFuture<Void> future = env.adminClient().createTopics(
+                Collections.singleton(new NewTopic("myTopic", Collections.singletonMap(Integer.valueOf(0), asList(new Integer[]{0, 1, 2})))),
```
nit: this can be simplified `Collections.singletonMap(0, asList(0, 1, 2)`

A few more of these in this file.

cmccabe committed:
KAFKA-6299. Fix AdminClient error handling when metadata changes

When AdminClient gets a NOT_CONTROLLER error, it should refresh its
metadata and retry the request, rather than making the end-user deal
with NotControllerException.

Move AdminClient's metadata management outside of NetworkClient and into
AdminMetadataManager.  This will make it easier to do more sophisticated
metadata management in the future, such as implementing a NodeProvider
which fetches the leaders for topics.

Rather than manipulating newCalls directly, the AdminClient service
thread now drains it directly into pendingCalls.  This minimizes the
amount of locking we have to do, since pendingCalls is only accessed
from the service thread.

cmccabe commented:
Jenkins is flaking again due to out of memory errors launching git.
```
18:28:10 Caused by: java.lang.OutOfMemoryError: unable to create new native thread
18:28:10 	at java.lang.Thread.start0(Native Method)
18:28:10 	at java.lang.Thread.start(Thread.java:717)
18:28:10 	at hudson.Proc$LocalProc.<init>(Proc.java:269)
18:28:10 	at hudson.Proc$LocalProc.<init>(Proc.java:218)
18:28:10 	at hudson.Launcher$LocalLauncher.launch(Launcher.java:930)
18:28:10 	at hudson.Launcher$ProcStarter.start(Launcher.java:450)
18:28:10 	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:1992)
18:28:10 	... 15 more
18:28:10 ERROR: Error cloning remote repo 'origin'
18:28:10 Retrying after 10 seconds
```

cmccabe commented:
retest this please

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/AdminMetadataManager.java:
```diff
@@ -0,0 +1,248 @@
+ * limitations under the License.
+ */
+
+package org.apache.kafka.clients.admin;
```
This is a public package, which conventionally has implied that all of the classes in it are public as well. Should we have an `internals` package for stuff like this? This is the pattern we use for the consumer and producer.

cmccabe replied:
OK, I'll create an internals package, to follow the convention.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/AdminMetadataManager.java:
```diff
@@ -0,0 +1,248 @@
+ */
+public class AdminMetadataManager {
+    /**
+     * The slf4 logger.
```
nit: comments like this seem like overkill

cmccabe replied:
fair enough

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/AdminMetadataManager.java:
```diff
@@ -0,0 +1,248 @@
+     */
+    private AuthenticationException authException = null;
+
+    public class AdminMetadataUpdater implements MetadataUpdater {
```
Would it make sense to extend `ManualMetadataUpdater`. It already has most of the no-op functionality we want.

cmccabe replied:
I guess my thought process here is that the broker is using ManualMetadataUpdater, and I don't want changes to ManualMetadataUpdater to change AdminClient.  Since the (lack of?) functionality here is minimal, seems better just to create another class.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -491,23 +515,43 @@ public Node provide() {
+        }
 
         Call(String callName, long deadlineMs, NodeProvider nodeProvider) {
+            this.internal = false;
```
nit: use the other constructor?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/AdminMetadataManager.java:
```diff
@@ -0,0 +1,248 @@
+        @Override
+        public void handleAuthenticationFailure(AuthenticationException e) {
+            log.info("AdminMetadataManager got AuthenticationException", e);
+            update(Cluster.empty(), time.milliseconds(), e);
```
It's a little unclear if we need this. Below when we see a disconnect, we check for authentication errors explicitly. Do we need anything else? It would be a little clearer if we only have one path for surfacing authentication errors.

cmccabe replied:
When NetworkClient gets a request, it sometimes has to make additional internal requests to fulfill it.  For example, if NC gets asked to make an AlterConfigsRequest, it may first have to make an API versions request.  However, if this API versions request fails, there is nothing which ties the failure back to the AlterConfigsRequest.  Since it's an "internal" request, the failure disappears without a trace and never makes its way into the list of responses.

I would argue that this is a bug in NetworkClient.  We are currently hacking around it by things like having the event loop manually iterate through each node in ClusterConnectionStates to see whether any of them ended up in ConnectionState.AUTHENTICATION_FAILED.   The metadata updater is another hack which gets triggered even when the Response gets dropped, in NetworkClient#processDisconnection.

If we want to be a little braver, we could say that when handling an internal APIVersionRequest disconnect, we could also send a disconnection to the next non-internal request queued for that node.

Really the whole concept of an internal request is evil.  We should just have a wrapper class around NetworkClient that translates one stream of requests into another.  That would help us keep this straight.  But that is too big to do right now.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -970,54 +968,105 @@ private void handleResponses(long now,
         @Override
         public void run() {
-            /*
+            /**
```
nit: it's a bit weird to use javadoc comments for local variables... Also, I would have expected implicitly that they are all only accessed from this thread. No harm emphasizing it I guess, but it makes me look around for the state the variables that _are_ shared with other threads.

cmccabe replied:
I think the JavaDoc is useful here.  And you can never be too explicit about concurrency ;)

hachikuji replied:
Can local variables even be javadoc'd? I think regular comments would be fine.

Sure, no harm in being explicit about concurrency, but my default assumption for a local variable is that it would not be shared across threads, so the doc felt like overkill.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/Cluster.java:
```diff
@@ -274,7 +274,7 @@ public Node controller() {
     @Override
     public String toString() {
-        return "Cluster(id = " + clusterResource.clusterId() + ", nodes = " + this.nodes + ", partitions = " + this.partitionsByTopicPartition.values() + ")";
+        return "Cluster(id = " + clusterResource.clusterId() + ", nodes = " + this.nodes + ", partitions = " + this.partitionsByTopicPartition.values() + ", controller = " + controller + ")";
```
nit: split to multiple lines

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/MockClient.java:
```diff
@@ -174,6 +180,26 @@ public void disconnect(String node) {
+                authErrorIter.remove();
+                // Set up a disconnected ClientResponse and create an authentication error
+                // for the affected node.
+                authenticationFailed(node, timeoutMs);
```
Is this not a concurrent modification?

cmccabe replied:
Technically, the iterator removes the entry from `pendingAuthenticationErrors` via `Iterator#remove`, and then the call to `pendingAuthenticationErrors#remove` has no effect.

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/MockClient.java:
```diff
@@ -128,11 +129,16 @@ public void blackout(Node node, long duration) {
         blackout(node, duration);
     }
 
+    public void createPendingAuthenticationError(Node node, long timeoutMs) {
```
The name `timeoutMs` seems misleading. It's really the blackout period following the authentication error?

cmccabe replied:
Good point.  I will change it to `blackoutMs`.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -1106,6 +1163,40 @@ void call(Call call, long now) {
+                    if (e instanceof AuthenticationException) {
+                        log.info("Unable to fetch cluster metadata from node {} because of " +
+                            "authentication error", curNode(), e);
+                        metadataManager.update(Cluster.empty(), time.milliseconds(), (AuthenticationException) e);
```
What is the expected behavior if the user ignores the auth error and continues to use the AdminClient?

cmccabe replied:
It will continue to throw `AuthenticationException`.

After enough time, another metadata request may be made which may succeed, which would allow future requests to go through.  But we don't spam metadata requests or anything-- if the auth exception is cleared, it will be because of a timeout.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/AdminMetadataManager.java:
```diff
@@ -0,0 +1,248 @@
+
+    void clearController() {
+        if (cluster.controller() != null) {
+            if (log.isTraceEnabled()) {
```
nit: not saving much..

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -1026,13 +1075,19 @@ public void run() {
+                    chooseNodesForPendingCalls(now, pendingCalls.listIterator(pendingCalls.size() - 1),
+                        callsToSend);
                 }
+                pollTimeout = Math.min(pollTimeout,
```
Should the poll timeout take into account the time to the next metadata refresh? What happens if we are in the middle of the metadata backoff when a call is made?

cmccabe replied:
Yes, it should.  Will fix.

hachikuji reviewed (commented):
Thanks, the approach using the custom metadata updater seems promising. Left some comments/questions.

cmccabe committed:
Fixes

cmccabe commented:
retest this please

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -698,43 +744,15 @@ int nextTimeoutMs() {
          */
-        private synchronized void timeoutNewCalls(TimeoutProcessor processor) {
-            int numTimedOut = processor.handleTimeouts(newCalls,
+        private synchronized void timeoutPendingCalls(TimeoutProcessor processor, List<Call> pendingCalls) {
```
Is the synchronization needed here since `pendingCalls` is only accessed by this thread?

cmccabe replied:
It's not needed.  Good catch.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -743,56 +761,60 @@ private synchronized void timeoutNewCalls(TimeoutProcessor processor) {
+                    node = call.nodeProvider.provide();
+                } catch (Throwable t) {
+                    // Handle authentication errors while choosing nodes.
+                    log.debug("Unable to choose node for {}: {}", call, t);
```
We don't need the second `{}` since the logging treats the exception specially.

cmccabe committed:
More review fixes

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/internal/AdminMetadataManager.java:
```diff
@@ -0,0 +1,247 @@
+
+    public boolean isReady() {
+        if (authException != null) {
+            log.trace("Metadata is ready: got authentication exception.");
```
Should this be "Metadata is **not** ready"?

cmccabe committed:
Fix TopicAdminTest to set a node to be returned from leastLoadedNodes

cmccabe committed:
Change log message in AdminMetadataManager#isReady

hachikuji reviewed (approved):
Thanks for the patch, LGTM. Will merge after the builds complete.

hachikuji commented:
retest this please
